File: javascript/leetcode/33. Search in Rotated Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  let start = 0;
  let end = nums.length - 1;

  while (start <= end) {
    let mid = Math.floor((start + end) / 2);
    let val = nums[mid];

    if (target === val) return mid;

    if (val > nums[end]) {
      target < val && target >= nums[start] ? (end = --mid) : (start = ++mid);
    } else if (val < nums[start]) {
      target > val && target <= nums[end] ? (start = ++mid) : (end = --mid);
    } else {
      target < val ? (end = --mid) : (start = ++mid);
    }
  }

  return -1;
};

File: c++/leetcode/33. Search in Rotated Sorted Array.cpp
class Solution
{
public:
  int search(vector<int> &nums, int target)
  {
    int start = 0;
    int end = nums.size() - 1;

    while (start <= end)
    {
      int mid = (start + end) / 2;
      int val = nums[mid];

      if (val == target)
        return mid;

      if (val > nums[end])
      {
        if (target < val && target >= nums[start])
          end = --mid;
        else
          start = ++mid;
      }

      else if (val < nums[start])
      {
        if (target > val && target <= nums[end])
          start = ++mid;
        else
          end = --mid;
      }

      else
      {
        if (target < val)
          end = --mid;
        else
          start = ++mid;
      }
    }

    return -1;
  }
};